app/controllers/application_controller.rb

```diff
@@ -40,15 +40,25 @@ def configure_permitted_parameters
     end
   end
 
+  # Check user's password against pwned password service and warn if necessary
+  def check_for_pwnage( resource )
+    return unless resource.respond_to?( :pwned? ) && resource.pwned?
+
+    # :nocov:
+    set_flash_message! :alert, :warn_pwned
+    # :nocov:
+  end
+
+  # Override post-login redirect
   def after_sign_in_path_for( resource )
-    # Check user's password against pwned password service and warn if necessary
-    if resource.respond_to?( :pwned? ) && resource.pwned?
-      # :nocov:
-      set_flash_message! :alert, :warn_pwned
-      # :nocov:
+    check_for_pwnage( resource )
+
+    if request.referer.present? && request.referer != new_user_session_url
+      return request.referer
```

Comment:
I'm not sure, but there may be a potential security vulnerability here, if the referer is a 3rd-party (attacker-owned) site.

Perhaps if combined with other vulnerabilities, ShinyCMS could be made to redirect a user to a phishing copy site pretending to be the genuine site.

Consider only returning the referer path here, removing any host. 

~~For extra confidence, consider checking the referer path against the app's known `GET` routes using `Rails.application.routes.recognize_path`.~~ <- Safelist probably overkill now I think about it, but returning only the path without the host is worthwhile.